docker: add a patch to fix Docker buildkit when using ZFS graph driver.

jaen opened the pull request:
###### Description of changes

Buildkit had problems building Docker images when using the ZFS graph driver. A fix suggestion was made basing on a similar issue with the BTRFS driver and it worked. The patch incorporates changes merged into the upstream in this PR:
https://github.com/moby/moby/pull/43136

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux (NOTE: via qemu binfmt wrapper)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
  NOTE: I have been using Docker with this patch for a while and it worked well, but technically it wasn't from this commit, but via an override (that shouldn't matter thought, right?)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jaen committed:
docker: add a patch to fix Docker buildkit when using ZFS graph driver.

The patch incorporates changes merged into the upstream in this PR:
https://github.com/moby/moby/pull/43136

mikroskeem reviewed (approved):
I have experienced this issue myself, but never thought about pointing finger towards moby. LGTM

vdemeester reviewed (approved):
LGTM 🐯

jaen commented:
Any interest in merging it (I can't do this)? I don't really mind carrying this patch in my configurations, but could be useful for other people until 21.y.z releases.

maxeaubrey commented:
@jaen Could you change the PR to use `fetchpatch` instead (either on the PR or commit) to make it easier to track this, as well as change the comment to reference the PR? Then it's good to go from my view.

I'm generally wary of bringing in patches into software like this, but given how common using ZFS on nix is, I'm not opposed here.

jaen commented:
@maxeaubrey but the comment already refers to the Docker PR? Or do you mean I should reference this PR here instead? Could you clarify please? Otherwise I'll fix it up in a minute.

jaen commented:
Ok, changed to `fetchpatch` with the PR as the patch source (so I dropped linking the PR from the comment again). The PR ix also linked in the commit message. I have also rebased on top of master.

If this is not what you wanted, then please do let me know what changes you require.

maxeaubrey commented:
Sorry, I must have missed there being a link there. This looks good, thanks!
